File: app/src/main/java/demos/view/FitTextView.java
package demos.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.widget.TextView;

import java.util.List;

/**
 * Created by Victor on 16/5/10.
 */
public class FitTextView extends TextView {

    private Paint textPaint;
    private float eachWidth;
    private float spacingWidth;
    private float currX;
    private String textStr;

    public FitTextView(Context context) {
        super(context);
        init();
    }

    public FitTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public static void adjustTextViewsWidth(List<FitTextView> textViews) {
        adjustTextViewsWidth(textViews.toArray(new FitTextView[]{}));
    }

    public static void adjustTextViewsWidth(FitTextView... textViews) {
        int maxWidth = 0;
        for (FitTextView textView : textViews) {
            maxWidth = textView.getTextWidth() > maxWidth ? textView.getTextWidth() : maxWidth;
        }
        for (FitTextView textView : textViews) {
            textView.adjustWidth(maxWidth);
        }
    }

    private void init() {
        textStr = getText().toString();
        textPaint = new Paint();
        textPaint.setTextSize(getTextSize());
        textPaint.setAntiAlias(true);
        textPaint.setColor(getCurrentTextColor());
        textPaint.setTextAlign(Paint.Align.LEFT);
        setMinWidth((int) textPaint.measureText(textStr));
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        measureText();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        currX = 0f;
        for (int i = 0; i < textStr.length(); i++) {
            canvas.drawText(String.valueOf(textStr.charAt(i)), currX, getBaseline(), textPaint);
            currX += eachWidth + spacingWidth;
        }
    }

    public String getRealText() {
        return textStr;
    }

    public void setText(String text) {
        textStr = text;
        measureText();
        postInvalidate();
    }

    private void measureText() {
        if (textPaint.measureText(textStr) > getMinWidth()) {
            setMinWidth((int) textPaint.measureText(textStr));
        }
        eachWidth = textPaint.measureText(textStr) / textStr.length();
        spacingWidth = (getMeasuredWidth() - textPaint.measureText(textStr)) / (textStr.length() - 1);
    }

    private void adjustWidth(int width) {
        setMinWidth(width);
        measureText();
        postInvalidate();
    }

    private int getTextWidth() {
        return textPaint.measureText(textStr) > getMinWidth() ? (int) textPaint.measureText(textStr) : getMinWidth();
    }
}

File: app/src/main/kotlin/demos/view/FitTextView.kt
package demos.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.widget.TextView

/**
* Author : Victor
* Time : 16/5/10
*/
class FitTextView : TextView {
    companion object {
        fun adjustTextViewsWidth(textViews: List<FitTextView>) {
            var maxWidth = 0
            textViews.forEach { maxWidth = if (it.getTextWidth() > maxWidth) it.getTextWidth() else maxWidth }
            textViews.forEach { it.adjustWidth(maxWidth) }
        }
    }

    private var textPaint: Paint
    private var eachWidth = 0f
    private var spacingWidth = 0f
    private var currX = 0f
    private var textStr: String

    constructor(context: Context, attributeSet: AttributeSet) : super(context, attributeSet, 0) {
        textStr = text.toString()
        textPaint = Paint()
        textPaint.textSize = textSize
        textPaint.isAntiAlias = true
        textPaint.color = currentTextColor
        textPaint.textAlign = Paint.Align.LEFT
        minWidth = textPaint.measureText(textStr).toInt()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        measureText()
    }

    override fun onDraw(canvas: Canvas) {
        currX = 0f
        for (i in 0..textStr.length - 1) {
            canvas.drawText(textStr.elementAt(i).toString(), currX, baseline.toFloat(), textPaint)
            currX += eachWidth + spacingWidth
        }
    }

    fun getRealText() = textStr

    fun setText(text: String) {
        textStr = text
        measureText()
        postInvalidate()
    }

    private fun measureText() {
        if (textPaint.measureText(textStr) > minWidth) {
            minWidth = textPaint.measureText(textStr).toInt()
        }
        eachWidth = textPaint.measureText(textStr) / textStr.length
        spacingWidth = (measuredWidth - textPaint.measureText(textStr)) / (textStr.length - 1)
    }

    private fun adjustWidth(width: Int) {
        minWidth = width
        measureText()
        postInvalidate()
    }

    fun getTextWidth() = if (textPaint.measureText(textStr) > minWidth) textPaint.measureText(textStr).toInt() else minWidth
}